src/services/organizeImports.ts

```diff
@@ -492,4 +491,14 @@ namespace ts.OrganizeImports {
                 return 6;
         }
     }
+
+    function getNewImportSpecifiers(namedImports: ImportDeclaration[]) {
+        return flatMap(namedImports, namedImport =>
+            map((namedImport.importClause?.namedBindings as NamedImports).elements, importSpecifier =>

```

Comment:
Should this still be `namedImport.importClause!.namedBindings`? The way it's currently been rewritten, if `importClause` is `undefined`, the access to `elements` will break at runtime, so it seems like `!` was fine.

Reply:
`map` accepts [] | undefined

Reply:
https://github.com/microsoft/TypeScript/blob/2172e1964f6e03a4912c11a7c133f3b95d86ccb4/src/compiler/core.ts#L308-L319

https://github.com/microsoft/TypeScript/blob/2172e1964f6e03a4912c11a7c133f3b95d86ccb4/src/compiler/core.ts#L380-L396

> ...if importClause is undefined

`map` returns `undefined`, and `flatMap` returns `empty array`

Reply:
What I mean is that in the following code

```
(namedImport.importClause?.namedBindings as NamedImports).elements
```

if the assertion is wrong, then this has the potential to crash before the call to `map`, because you'll end up running `undefined.elements`.

But it was already similarly the case for the old code:

```
(namedImport.importClause!.namedBindings as NamedImports).elements
```

where you'd end up running `undefined.namedBindings`.

So I guess I don't understand the need to switch to a `?.`.

Maybe it's unclear, but [a parenthesized type assertion *does* break the optional chain](https://www.typescriptlang.org/play?#code/DYUwLgBAHhC8EAoCGB+AdAIwkgztgdgJ4CUaAxkA).

Reply:
I've moved this logic to `tryGetNamedBindingElements`. Is it ok?